compiler/rustc_mir/src/transform/pass.rs

```diff
@@ -0,0 +1,151 @@
+use std::borrow::Cow;
+
+use rustc_middle::mir::{Body, MirPhase};
+use rustc_middle::ty::TyCtxt;
+
+use super::{dump_mir, validate};
+
+/// A streamlined trait that you can implement to create a pass; the
+/// pass will be named after the type, and it will consist of a main
+/// loop that goes over each available MIR and applies `run_pass`.
+pub trait MirPass<'tcx> {
+    const LEVEL: OptLevel;
+
+    fn name(&self) -> Cow<'_, str> {
+        let name = std::any::type_name::<Self>();
+        name.rsplit(':').next().unwrap().into()
+    }
+
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>);
+}
+
+const UNSOUND_MIR_OPT_LEVEL: u8 = 2;
+
+pub enum OptLevel {
+    /// Passes that will run at `-Zmir-opt-level=N` or higher.
+    N(u8),
+
+    /// Passes that are known or suspected to cause miscompilations.
+    ///
+    /// These passes only run if `-Zunsound-mir-opts` is enabled and we are at `mir-opt-level=2` or
+    /// above.
+    Unsound,
+
+    /// Passes that clean up the MIR after other transformations.
+    ///
+    /// A `Cleanup` pass is skipped if the pass immediately preceding it is skipped.
+    //
+    // FIXME: Maybe we want to run the cleanup unless *all* passes between it and the last cleanup
+    // were skipped?
```

Comment:
This seems a bit fragile to me. Several optimizations nowadays don't need a cleanup after them, because they just invoke the appropriate cleanup method within themselves. Maybe we should migrate to that scheme in general and get rid of all the cleanup passes? That could make diffs harder to read though, so not sure. It does make diffs harder to read in some of the passes with integrated cleanup

Reply:
Yeah, I don't really have a good solution for this. The status quo is equivalent to giving each instance of a cleanup pass a different `LEVEL`. I could do this with const-generics, or I could switch from an associated const to an instance method. The latter would make the `MirPass` trait object-safe again, meaning I could get rid of the `run_passes` macro, but I really like the declarative nature of associated constants.

Reply:
Thinking about it again, to preserve the current situation, only running a cleanup if the MIR wasn't modified since the last *same* cleanup doesn't seem too bad. Approximating "MIR wasn't modified" with "no MIR passes were run" is also ok.

The FIXME should be resolved before this PR is merged, as otherwise we may be missing cleanups that were previously triggered even if there were multiple other not-running optimizations in between.